src/Resource/Resource/Collection.js

```diff
@@ -9,111 +11,131 @@ function Collection(name, repository) {
     throw Error('Collection() must be called with new');
   }
 
+  this._repository = repository;
+
   Object.defineProperties(this, {
-    name: {
-      enumerable: true,
-      configurable: false,
-      writable: false,
-      value: name,
-    },
-    repository: {
-      enumerable: true,
-      configurable: false,
-      writable: false,
-      value: repository,
-    },
-    metadata: {
+    _state: {
       enumerable: false,
       configurable: false,
+      writable: true,
       value: {
-        loading: false,
-        loaded: false,
-        parameterBag: null,
-        lastParameterBagState: null,
-        paging: {
-          count: null,
-          pages: null,
+        name,
+        data: {
+          items: [],
+        },
+        metadata: {
+          loading: false,
+          loaded: false,
+          parameterBag: null,
+          lastParameterBagState: null,
+          paging: {
+            count: null,
+            pages: null,
+          },
         },
-      },
-    },
-    data: {
-      enumerable: false,
-      configurable: false,
-      value: {
-        items: [],
       },
     },
   });
 }
 
 Collection.prototype = {
+  get name() {
+    return this._state.name;
+  },
+
+  get repository() {
+    return this._repository;
+  },
+
   get parameterBag() {
-    if (!this.metadata.parameterBag) {
-      this.metadata.parameterBag = new ParameterBag();
+    const parameterBag = new ParameterBag();
+    if (!this._state.metadata.parameterBag) {
```

Comment:
Kan dit niet ook gewon een ParameterBag.fromState(this._state.metadata.parameterBag || {}) zijn?

ParameterBag.fromState kan dan zijn:

ParameterBag.fromState(state) {
const parameterBag = new ParameterBag();
parameterBag.state = Object.assign(parameterBag.state, state);
return parameterBag;
}

Reply:
Goed idee!